src/app/InteractionModelEngine.cpp

```diff
@@ -530,5 +530,125 @@ bool InteractionModelEngine::IsOverlappedAttributePath(ClusterInfo & aAttributeP
     return false;
 }
 
+void InteractionModelEngine::DispatchCommand(CommandHandler & apCommandObj, const ConcreteCommandPath & aCommandPath,
+                                             TLV::TLVReader & apPayload)
+{
+    CommandHandlerInterface * handler = FindCommandHandler(aCommandPath.mEndpointId, aCommandPath.mClusterId);
+
+    if (handler)
+    {
+        CommandHandlerInterface::HandlerContext context(apCommandObj, aCommandPath, apPayload);
+        handler->InvokeCommand(context);
+
+        //
+        // If the command was handled, don't proceed any further and return successfully.
+        //
+        if (context.mCommandHandled)
+        {
+            return;
+        }
+    }
+
+    DispatchSingleClusterCommand(aCommandPath, apPayload, &apCommandObj);
+}
+
+bool InteractionModelEngine::CommandExists(const ConcreteCommandPath & aCommandPath)
+{
+    CommandHandlerInterface * handler = FindCommandHandler(aCommandPath.mEndpointId, aCommandPath.mClusterId);
+    if (handler)
+    {
+        return true;
+    }
+
+    //
+    // If we couldn't find a matching handler, default to querying the application using the following method
+    // to deduce if a command exists.
+    //
+    return ServerClusterCommandExists(aCommandPath);
+}
+
+CHIP_ERROR InteractionModelEngine::RegisterCommandHandler(CommandHandlerInterface * handler)
+{
+    VerifyOrReturnError(handler != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    for (auto * cur = mpCommandHandlerList; cur; cur = cur->GetNext())
+    {
+        if (cur->Matches(*handler))
+        {
+            ChipLogError(InteractionModel, "Duplicate command handler registration failed");
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+    }
+
+    handler->SetNext(mpCommandHandlerList);
+    mpCommandHandlerList = handler;
+
+    return CHIP_NO_ERROR;
+}
+
+void InteractionModelEngine::DeregisterCommandHandlers(EndpointId endpointId)
+{
+    CommandHandlerInterface * prev = nullptr;
+
+    for (auto * cur = mpCommandHandlerList; cur; cur = cur->GetNext())
+    {
+        if (cur->MatchesExactly(endpointId))
+        {
+            if (prev == nullptr)
+            {
+                mpCommandHandlerList = cur->GetNext();
+            }
+            else
+            {
+                prev->SetNext(cur->GetNext());
+            }
+        }
```

Comment:
Need to `cur->SetNext(nullptr)` if we removed `cur` from the list.

Reply:
Sure. I'll make the requisite fix to the equivalent logic in `attribute-storage.cpp` for the attribute access interface as well.